frontend_server/lib/server.dart

```diff
@@ -0,0 +1,160 @@
+import 'package:kernel/target/flutter_fasta.dart';
+import 'package:kernel/target/targets.dart';
+import 'package:usage/uuid/uuid.dart';
+
+ArgParser argParser = new ArgParser(allowTrailingOptions: true)
+  ..addFlag('train',
+      help: 'Run through sample command line to produce snapshot',
+      negatable: false)
+  ..addFlag('incremental',
+      help: 'Run compiler in incremental mode', defaultsTo: false)
+  ..addOption('sdk-root',
+      help: 'Path to sdk root',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk')
+  ..addOption('platform-kernel-dill',
+      help: 'Path to platform kernel dill file',
+      defaultsTo: '../../out/android_debug/flutter_patched_sdk/platform.dill');
+
+String usage = '''
+Usage: server [options] [input.dart]
+
+If input dart source code is provided on the command line, then the server 
+compiles it, generates dill file and exits.
+If no input dart source is provided on the command line, server waits for 
+instructions from stdin.
+
+Instructions:
+- compile <input.dart>
+- recompile <boundary-key>
+<invalidate-file1.dart>
+<invalidate-file2.dart>
+...
+<boundary-key>
+- accept
+- reject
+- quit
+
+Output:
+- result <boundary-key>
+<compiler output>
+<boundary-key> [<output.dill>]
+
+Options:
+${argParser.usage}
+''';
+
+enum State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
+
+starter(List<String> args) async {
+  ArgResults options = argParser.parse(args);
+  if (options['train']) {
+    return 0;
+  }
+
+  if (options.rest.isNotEmpty) {
+    await compile(options.rest[0], options);
+    return 0;
+  }
+
+  State state = State.READY_FOR_INSTRUCTION;
+  String boundaryKey;
+  print('Frontend server is ready.');
+  stdin
+      .transform(UTF8.decoder)
+      .transform(new LineSplitter())
+      .listen((String string) async {
+    switch (state) {
+      case State.READY_FOR_INSTRUCTION:
+        if (string.startsWith('compile ')) {
+          String filename = string.substring('compile '.length);
+          await compile(filename, options);
+        } else if (string.startsWith('recompile ')) {
+          boundaryKey = string.substring('recompile '.length);
+          state = State.RECOMPILE_LIST;
+        } else if (string == 'accept') {
+          ikg.acceptLastDelta();
+        } else if (string == 'reject') {
+          ikg.rejectLastDelta();
+        } else if (string == 'quit') {
+          exit(0);
+        }
+        break;
+      case State.RECOMPILE_LIST:
+        if (string == boundaryKey) {
+          recompileDelta();
+          state = State.READY_FOR_INSTRUCTION;
+        } else {
+          ikg.invalidate(Uri.parse(string));
+        }
+        break;
+    }
+  });
+}
+
+IncrementalKernelGenerator ikg;
+String filename;
+
+compile(String filename, ArgResults options) async {
+  var program;
+  String boundaryKey = new Uuid().generateV4();
+  print("result $boundaryKey");
+  if (options['incremental']) {
+    final CompilerOptions compilerOptions = new CompilerOptions()
```

Comment:
> One thing makes me uneasy is check and append of forward slash to the path. I'm not sure how this will work on Windows.

Good point - in fact we use uri everywhere in the compiler to avoid those issues. I believe in this case the Uri.base.resolve you do immediately after will ensure all slashes become `/`. Alternatively, you can include it after you resolve it as a uri.

We actually do the same normalization internally in `ProcessedOptions` (which is derived from `CompilerOptions`), so we don't depend on this in the FE. The main reason it may be important is that if you don't have it, then the call further below to resolve `platform.dill` will not work correctly.

> What is the issue XYZ that you are referring to above?
The issue, not sure if we have a preciseone, the closest one I can think of is http://github.com/dart-lang/sdk/issues/30111 ?